File: rust/src/executor.rs
//! Execute commands using tokio

use std::process::Stdio;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::{Child, Command};

/// Command execution result
#[derive(Debug, Clone, Default)]
pub struct ExecutionResult {
    pub exit_code: i32,
    pub stdout: String,
    pub stderr: String,
    pub command: String,
}

/// Execute a command and return the result
///
/// # Arguments
/// * `command` - Command to execute
/// * `dry_run` - If true, just return the command without executing
/// * `attached` - If true, stream output to console
///
/// # Returns
/// Execution result
pub async fn execute_command(
    command: &str,
    dry_run: bool,
    attached: bool,
) -> Result<ExecutionResult, std::io::Error> {
    if dry_run {
        println!("Dry run - command that would be executed:");
        println!("{}", command);
        return Ok(ExecutionResult {
            exit_code: 0,
            stdout: String::new(),
            stderr: String::new(),
            command: command.to_string(),
        });
    }

    let mut child = Command::new("bash")
        .arg("-c")
        .arg(command)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()?;

    let mut stdout = String::new();
    let mut stderr = String::new();

    // Read stdout
    if let Some(stdout_pipe) = child.stdout.take() {
        let mut reader = BufReader::new(stdout_pipe).lines();
        while let Some(line) = reader.next_line().await? {
            stdout.push_str(&line);
            stdout.push('\n');
            if attached {
                println!("{}", line);
            }
        }
    }

    // Read stderr
    if let Some(stderr_pipe) = child.stderr.take() {
        let mut reader = BufReader::new(stderr_pipe).lines();
        while let Some(line) = reader.next_line().await? {
            stderr.push_str(&line);
            stderr.push('\n');
            if attached {
                eprintln!("{}", line);
            }
        }
    }

    let status = child.wait().await?;
    let exit_code = status.code().unwrap_or(1);

    Ok(ExecutionResult {
        exit_code,
        stdout,
        stderr,
        command: command.to_string(),
    })
}

/// Process handle for non-blocking command execution
pub struct ProcessHandle {
    pub command: String,
    child: Option<Child>,
    stdout: String,
    stderr: String,
    exit_code: Option<i32>,
}

impl ProcessHandle {
    /// Create a new process handle
    fn new(command: String, child: Child) -> Self {
        Self {
            command,
            child: Some(child),
            stdout: String::new(),
            stderr: String::new(),
            exit_code: None,
        }
    }

    /// Wait for the process to exit
    pub async fn wait_for_exit(&mut self) -> Result<i32, std::io::Error> {
        if let Some(exit_code) = self.exit_code {
            return Ok(exit_code);
        }

        if let Some(mut child) = self.child.take() {
            // Read remaining stdout
            if let Some(stdout_pipe) = child.stdout.take() {
                let mut reader = BufReader::new(stdout_pipe).lines();
                while let Some(line) = reader.next_line().await? {
                    self.stdout.push_str(&line);
                    self.stdout.push('\n');
                }
            }

            // Read remaining stderr
            if let Some(stderr_pipe) = child.stderr.take() {
                let mut reader = BufReader::new(stderr_pipe).lines();
                while let Some(line) = reader.next_line().await? {
                    self.stderr.push_str(&line);
                    self.stderr.push('\n');
                }
            }

            let status = child.wait().await?;
            self.exit_code = Some(status.code().unwrap_or(1));
        }

        Ok(self.exit_code.unwrap_or(1))
    }

    /// Get collected output
    pub fn get_output(&self) -> (&str, &str, Option<i32>) {
        (&self.stdout, &self.stderr, self.exit_code)
    }

    /// Check if process has exited
    pub fn has_exited(&self) -> bool {
        self.exit_code.is_some()
    }
}

/// Start a command execution without waiting for completion
///
/// # Arguments
/// * `command` - Command to execute
/// * `attached` - If true, stream output to console
///
/// # Returns
/// Process handle
pub async fn start_command(
    command: &str,
    _attached: bool,
) -> Result<ProcessHandle, std::io::Error> {
    let child = Command::new("bash")
        .arg("-c")
        .arg(command)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()?;

    Ok(ProcessHandle::new(command.to_string(), child))
}

/// Execute a command in the background (detached)
///
/// # Arguments
/// * `command` - Command to execute
///
/// # Returns
/// Process ID if available
pub async fn execute_detached(command: &str) -> Result<Option<u32>, std::io::Error> {
    let child = Command::new("bash")
        .arg("-c")
        .arg(command)
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .stdin(Stdio::null())
        .spawn()?;

    Ok(child.id())
}

/// Signal handler cleanup function type
pub type CleanupFn = Box<dyn Fn() + Send + Sync>;

/// Setup CTRL+C handler for graceful shutdown
///
/// # Arguments
/// * `cleanup_fn` - Function to call on CTRL+C
///
/// # Returns
/// Function to remove the handler
pub fn setup_signal_handler<F>(cleanup_fn: F) -> impl Fn()
where
    F: Fn() + Send + Sync + 'static,
{
    // Note: In Rust with tokio, signal handling is typically done differently
    // This is a simplified version that uses ctrlc crate pattern
    // For production use, consider tokio::signal

    let cleanup = std::sync::Arc::new(cleanup_fn);
    let cleanup_clone = cleanup.clone();

    // Set up a simple flag for shutdown
    let shutdown = std::sync::Arc::new(std::sync::atomic::AtomicBool::new(false));
    let shutdown_clone = shutdown.clone();

    std::thread::spawn(move || {
        // This is a simplified pattern - in real code use tokio::signal
        loop {
            if shutdown_clone.load(std::sync::atomic::Ordering::Relaxed) {
                cleanup_clone();
                break;
            }
            std::thread::sleep(std::time::Duration::from_millis(100));
        }
    });

    move || {
        shutdown.store(true, std::sync::atomic::Ordering::Relaxed);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_execute_command_dry_run() {
        let result = execute_command("echo hello", true, false).await.unwrap();
        assert_eq!(result.exit_code, 0);
        assert_eq!(result.command, "echo hello");
    }

    #[tokio::test]
    async fn test_execute_command_real() {
        let result = execute_command("echo hello", false, false).await.unwrap();
        assert_eq!(result.exit_code, 0);
        assert!(result.stdout.contains("hello"));
    }

    #[tokio::test]
    async fn test_start_command_and_wait() {
        let mut handle = start_command("echo hello", false).await.unwrap();
        let exit_code = handle.wait_for_exit().await.unwrap();
        assert_eq!(exit_code, 0);

        let (stdout, _, _) = handle.get_output();
        assert!(stdout.contains("hello"));
    }

    #[tokio::test]
    async fn test_execute_detached() {
        let pid = execute_detached("sleep 0.1").await.unwrap();
        assert!(pid.is_some());
    }
}

File: js/src/executor.mjs
/**
 * Execute commands using the runtime's Node-compatible child_process API.
 */

/**
 * Execute a command and return the result
 * @param {string} command - Command to execute
 * @param {Object} [options] - Execution options
 * @param {boolean} [options.dryRun] - If true, just return the command without executing
 * @param {boolean} [options.attached] - If true, stream output to console
 * @param {Function} [options.onStdout] - Callback for stdout chunks
 * @param {Function} [options.onStderr] - Callback for stderr chunks
 * @param {Function} [options.onExit] - Callback for exit code
 * @returns {Promise<{exitCode: number, stdout: string, stderr: string}>} Execution result
 */
export async function executeCommand(command, options = {}) {
  const {
    dryRun = false,
    attached = true,
    onStdout,
    onStderr,
    onExit,
  } = options;

  if (dryRun) {
    console.log('Dry run - command that would be executed:');
    console.log(command);
    return { exitCode: 0, stdout: '', stderr: '', command };
  }

  const handle = await startCommand(command, {
    attached,
    onStdout,
    onStderr,
  });
  const exitCode = await handle.waitForExit();
  const { stdout, stderr } = handle.getOutput();

  if (onExit) {
    onExit(exitCode);
  }

  return { exitCode, stdout, stderr, command };
}

/**
 * Start a command execution without waiting for completion
 * @param {string} command - Command to execute
 * @param {Object} [options] - Execution options
 * @param {boolean} [options.attached] - If true, stream output to console
 * @param {boolean} [options.pipeStdin] - If true, open the child's stdin as a
 *   writable pipe (used by the per-command approval relay to send the prompt
 *   and permission responses as NDJSON frames)
 * @returns {Promise<Object>} Process handle with methods to interact with the process
 */
export async function startCommand(command, options = {}) {
  const { attached = true, onStdout, onStderr, pipeStdin = false } = options;
  const { spawn } = await import('node:child_process');

  let stdout = '';
  let stderr = '';
  let exitCode = null;
  let hasExited = false;
  let resolveExit;

  const exitPromise = new Promise((resolve) => {
    resolveExit = resolve;
  });

  const child = spawn('bash', ['-c', command], {
    stdio: [pipeStdin ? 'pipe' : 'ignore', 'pipe', 'pipe'],
  });

  child.stdout.on('data', (chunk) => {
    const data = chunk.toString();
    stdout += data;
    if (attached) {
      process.stdout.write(data);
    }
    if (onStdout) {
      onStdout(data);
    }
  });

  child.stderr.on('data', (chunk) => {
    const data = chunk.toString();
    stderr += data;
    if (attached) {
      process.stderr.write(data);
    }
    if (onStderr) {
      onStderr(data);
    }
  });

  child.on('error', (error) => {
    if (!hasExited) {
      stderr += error.message;
      if (attached) {
        process.stderr.write(error.message);
      }
      exitCode = 1;
      hasExited = true;
      resolveExit(1);
    }
  });

  child.on('close', (code) => {
    if (!hasExited) {
      exitCode = code ?? 0;
      hasExited = true;
      resolveExit(exitCode);
    }
  });

  // Give the stream a moment to start
  await Promise.resolve();

  return {
    command,
    waitForExit: () => exitPromise,
    getOutput: () => ({ stdout, stderr, exitCode, hasExited }),
    process: child,
    /**
     * Write a chunk to the child's stdin (no-op if stdin is not piped/open).
     * @param {string} data - Data to write
     * @returns {boolean} True when the write was attempted
     */
    writeStdin: (data) => {
      if (child.stdin && child.stdin.writable) {
        child.stdin.write(data);
        return true;
      }
      return false;
    },
    /**
     * Close the child's stdin (signals end-of-input).
     */
    endStdin: () => {
      if (child.stdin && !child.stdin.destroyed) {
        child.stdin.end();
      }
    },
  };
}

/**
 * Execute a command in the background (detached)
 * @param {string} command - Command to execute
 * @returns {Promise<{pid: number|null}>} Process information
 */
export async function executeDetached(command) {
  const { spawn } = await import('node:child_process');

  return new Promise((resolve, reject) => {
    try {
      const child = spawn('bash', ['-c', command], {
        detached: true,
        stdio: 'ignore',
      });

      child.unref();

      resolve({ pid: child.pid });
    } catch (error) {
      reject(error);
    }
  });
}

/**
 * Setup CTRL+C handler for graceful shutdown
 * @param {Function} cleanupFn - Function to call on CTRL+C
 * @returns {Function} Function to remove the handler
 */
export function setupSignalHandler(cleanupFn) {
  // Detect runtime
  const isDeno = typeof Deno !== 'undefined';
  const isNode =
    typeof process !== 'undefined' && process.versions && process.versions.node;

  if (isDeno) {
    // Deno signal handling
    const handler = async () => {
      console.log('\nReceived SIGINT, shutting down gracefully...');
      try {
        await cleanupFn();
      } catch (error) {
        console.error('Error during cleanup:', error.message);
      }
      Deno.exit(0);
    };

    // Listen for SIGINT in Deno
    Deno.addSignalListener('SIGINT', handler);

    return () => {
      try {
        Deno.removeSignalListener('SIGINT', handler);
      } catch {
        // Signal listener may already be removed
      }
    };
  } else if (isNode) {
    // Node.js signal handling
    const handler = async (signal) => {
      console.log(`\nReceived ${signal}, shutting down gracefully...`);
      try {
        await cleanupFn();
      } catch (error) {
        console.error('Error during cleanup:', error.message);
      }
      process.exit(0);
    };

    process.on('SIGINT', handler);
    process.on('SIGTERM', handler);

    return () => {
      process.off('SIGINT', handler);
      process.off('SIGTERM', handler);
    };
  }

  // Return no-op for other runtimes
  return () => {};
}
